lib/publish.js

```diff
@@ -1,167 +1,105 @@
+const pack = require('libnpmpack')
+const libpub = require('libnpmpublish').publish
+const runScript = require('@npmcli/run-script')
+
+const npm = require('./npm')
+const output = require('./utils/output')
+const otplease = require('./utils/otplease')
+const { getContents, logTar } = require('./utils/tar')
 
-publish.usage = 'npm publish [<tarball>|<folder>] [--tag <tag>] [--access <public|restricted>] [--dry-run]' +
+const readJson = util.promisify(require('read-package-json'))
+
+publish.usage = 'npm publish [<folder>] [--tag <tag>] [--access <public|restricted>] [--dry-run]' +
                 "\n\nPublishes '.' if no argument supplied" +
                 '\n\nSets tag `latest` if no --tag specified'
 
-publish.completion = function (opts, cb) {
-  // publish can complete to a folder with a package.json
-  // or a tarball, or a tarball url.
-  // for now, not yet implemented.
-  return cb()
-}
-
-const PublishConfig = figgyPudding({
+const publishConfig = () => ({
   dryRun: 'dry-run',
-  'dry-run': { default: false },
-  force: { default: false },
-  json: { default: false },
-  Promise: { default: () => Promise },
-  tag: { default: 'latest' },
-  tmp: {}
+  tag: 'defaultTag',
+  'dry-run': false,
+  defaultTag: 'latest',
+  json: false,
+  tmp: {},
+  ...npm.flatOptions
 })
 
 module.exports = publish
-function publish (args, isRetry, cb) {
-  if (typeof cb !== 'function') {
-    cb = isRetry
-    isRetry = false
-  }
+async function publish (args, cb) {
   if (args.length === 0) args = ['.']
   if (args.length !== 1) return cb(publish.usage)
 
-  log.verbose('publish', args)
+  log.verbose('publish', args) 
 
-  const opts = PublishConfig(npmConfig())
-  const t = opts.tag.trim()
+  const opts = publishConfig()
+  const t = opts.defaultTag.trim()
   if (semver.validRange(t)) {
     return cb(new Error('Tag name must not be a valid SemVer range: ' + t))
   }
 
-  return publish_(args[0], opts)
-    .then((tarball) => {
-      const silent = log.level === 'silent'
-      if (!silent && opts.json) {
-        output(JSON.stringify(tarball, null, 2))
-      } else if (!silent) {
-        output(`+ ${tarball.id}`)
-      }
-    })
-    .nodeify(cb)
-}
+  const tarball = await publish_(args[0], opts)
+  const silent = log.level === 'silent'
+  if (!silent && opts.json) {
+    output(JSON.stringify(tarball, null, 2))
+  } else if (!silent) {
+    output(`+ ${tarball.id}`)
+  }
 
-function publish_ (arg, opts) {
-  return statAsync(arg).then((stat) => {
-    if (stat.isDirectory()) {
-      return stat
-    } else {
-      const err = new Error('not a directory')
-      err.code = 'ENOTDIR'
-      throw err
-    }
-  }).then(() => {
-    return publishFromDirectory(arg, opts)
-  }, (err) => {
-    if (err.code !== 'ENOENT' && err.code !== 'ENOTDIR') {
-      throw err
-    } else {
-      return publishFromPackage(arg, opts)
-    }
-  })
+  return Promise.resolve(tarball)
+    .then(cb())
```

Comment:
```
.then(() => cb()).catch(cb)
```

Reply:
Actually, since this is an async function, you can just do `return tarball`.  It's automagically promisified.